keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,202 @@
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications. We will perform performance/benchmark tests on this change and backward compatibility test.
+
+ Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will abstract the `Node` and `Pod` label selection feature, and finally the `NodeSelector` and `PodSelector` will be based on the `LabelSelectorRequirement` implementation:
+
+ The definition of `NodeSelector` will not change,it is defined as below:
+ ```go
+type NodeSelector struct {
+  NodeSelectorTerms []NodeSelectorTerm
+}
+```
+
+For the definition of `NodeSelectorTerm`, today the api looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []NodeSelectorRequirement
+  MatchFields []NodeSelectorRequirement
+}
+```
+
+We will use `LabelSelectorRequirement` to replace the original `NodeSelectorRequirement` as the base selector of `NodeSelector`. So it looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []LabelSelectorRequirement
+  MatchFields []LabelSelectorRequirement
+}
+```
+
+`PodSelector` is our new selector. A pod selector represents the union of the results of one or more label queries over a set of pods. The of `PodSelector` is defined as below:
+ ```go
+type PodSelector struct {
+  MatchLabels map[string]string
+  MatchExpressions []LabelSelectorRequirement
+}
+```
+
+`LabelSelectorRequirement` is an abstract selector implementation of `NodeSelector` and `PodSelector`. And the `LabelSelectorRequirement` is defined as below:
```

Comment:
I am worried about this name, I think it needs to be more specific, since it is only appropriate in these two places, and not for use in general label selectors everywhere?

Reply:
It makes sense, I will change to `NodeAndPodSelectorRequirement`.